core/src/main/kotlin/io/github/kkafka/producer/transaction.kt

```diff
@@ -0,0 +1,91 @@
+package io.github.kkafka.producer
+
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.common.errors.TransactionAbortedException
+
+/**
+ * Produce transactionally with a given consumer on the given topics.
+ * The [consumer] indicates which consumer is taking part in the transaction with the producer.
+ * The [consumerTopics] indicate which of the consumer's topics are taking part in the transaction.
+ * You should not include the producer's topic.
+ * The [consumer] must be subscribed to and consuming from the given [consumerTopics].
+ *
+ * [block] is the transaction code. Here you can consume from [consumer] and send with the producer.
+ * At the end of the [block], if no exceptions were thrown, the transaction is committed and the result of the
+ * block is returned. Otherwise, the transaction is aborted and the exception is thrown.
+ *
+ * This is useful for a consume-process-produce pattern, where we commit to the consumer only if we have successfully
+ * produced. This ensures exactly once processing.
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(
+    consumer: KafkaConsumer<*, *>,
+    consumerTopics: Set<String>,
+    block: KafkaProducer<K, V>.() -> T,
+): T {
+    beginTransaction()
+    return try {
+        val res = block(this)
+
+        // Get the TopicPartition objects given the topic names
+        val consumerTopicPartitions = consumer.assignment().filter { it.topic() in consumerTopics }.map { it!! }
+
+        // Get the offsets of the partitions so we can commit them as part of the transaction
+        val offsets = consumerTopicPartitions.associateWith { OffsetAndMetadata(consumer.position(it)) }
+
+        val groupId = consumer.groupMetadata().groupId()
+
+        // Make sure that we got all the needed partitions
+        if (offsets.size != consumerTopics.size) {
+            throw TransactionAbortedException(
+                "consumer in transaction is not properly subscribed to all topics given for transaction"
+            )
+        }
+
+        // Commit the consumer's offsets
+        sendOffsetsToTransaction(offsets, groupId)
+
+        // Successful transaction
+        commitTransaction()
+
+        res
+    } catch (e: Exception) {
+        // If the transaction failed, abort
+        abortTransaction()
+        throw e
+    }
+}
+
+/**
+ * Consume and produce transactionally on a single topic.
+ * See [transaction].
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(
+    consumer: KafkaConsumer<*, *>,
+    consumerTopic: String,
+    block: KafkaProducer<K, V>.() -> T,
+): T {
+    return try {
```

Comment:
for single line returns, use expression syntax instead, like 

```kotlin
fun hello(): Int = 3
```

rather than

```kotlin
fun hello(): Int {
    return 3
}
```

it really does not matter but it's nice if it is consistent in codebase and linter does not catch it